### ISSUE TITLE
dist/Data-Dumper: blockers to strict-by-default

### ISSUE THREAD

**Issue by jkeenan** [atoomic/perl#223]
As of commit ac6b7ccf73 in the alpha-dev-02-strict branch on Aug 18 2020, 3 test files underneath `dist/Data-Dumper` are failing.  For each file, here is sample output from `cd t; ./perl harness -v <file>; cd -`:
```
# ../dist/Data-Dumper/t/bless.t

#   Failed test 'eval reverts dump'
#   at t/bless.t line 33.
#     Structures begin differing at:
#          $got = undef
#     $expected = a'b=HASH(0x55b519a67c68)

#   Failed test 'eval reverts dump'
#   at t/bless.t line 44.
#     Structures begin differing at:
#          $got = undef
#     $expected = a\=HASH(0x55b51a165430)

#   Failed test 'eval reverts dump'
#   at t/bless.t line 33.
#     Structures begin differing at:
#          $got = undef
#     $expected = a'b=HASH(0x55b51a1f5b80)

#   Failed test 'eval reverts dump'
#   at t/bless.t line 44.
#     Structures begin differing at:
#          $got = undef
#     $expected = a\=HASH(0x55b51a19a200)
# Looks like you failed 4 tests of 11.
../dist/Data-Dumper/t/bless.t .. 
1..11
ok 1 - use Data::Dumper;
# $Data::Dumper::Useperl = 0
ok 2 - package name in bless is escaped if needed
not ok 3 - eval reverts dump
ok 4 - package name in bless is escaped if needed
not ok 5 - eval reverts dump
ok 6 - We can dump blessed qr//'s properly
# $Data::Dumper::Useperl = 1
ok 7 - package name in bless is escaped if needed
not ok 8 - eval reverts dump
ok 9 - package name in bless is escaped if needed
not ok 10 - eval reverts dump
ok 11 - We can dump blessed qr//'s properly
Dubious, test returned 4 (wstat 1024, 0x400)
Failed 4/11 subtests 

Test Summary Report
-------------------
../dist/Data-Dumper/t/bless.t (Wstat: 1024 Tests: 11 Failed: 4)
  Failed tests:  3, 5, 8, 10
  Non-zero exit status: 4
Files=1, Tests=11,  0 wallclock secs ( 0.01 usr  0.00 sys +  0.05 cusr  0.00 csys =  0.06 CPU)
Result: FAIL
```
Perhaps a case where the problem lies in the fact that under strict-by-default, programs run by `eval STRING` must also be strict-by-default.
```
# ../dist/Data-Dumper/t/dumper.t

Global symbol "@c" requires explicit package name (did you forget to declare "my @c"?) at t/dumper.t line 147.
Global symbol "$c" requires explicit package name (did you forget to declare "my $c"?) at t/dumper.t line 148.
Global symbol "@c" requires explicit package name (did you forget to declare "my @c"?) at t/dumper.t line 148.
Global symbol "$c" requires explicit package name (did you forget to declare "my $c"?) at t/dumper.t line 150.
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at t/dumper.t line 308.
Global symbol "$d" requires explicit package name (did you forget to declare "my $d"?) at t/dumper.t line 309.
Global symbol "$foo" requires explicit package name (did you forget to declare "my $foo"?) at t/dumper.t line 376.
BEGIN not safe after errors--compilation aborted at t/dumper.t line 392.
../dist/Data-Dumper/t/dumper.t ..
Dubious, test returned 255 (wstat 65280, 0xff00)
No subtests run
```
I suspect that in the above I got part way through the file scoping those variables and then saw that I had created numerous unit test failures by doing so.
```
# ../dist/Data-Dumper/t/qr.t


#   Failed test 'Dumping (?^: \/ ) with XS'
#   at t/qr.t line 20.
# Global symbol "$VAR1" requires explicit package name (did you forget to declare "my $VAR1"?) at (eval 8) line 1.
# $VAR1 = qr/ \/ /;

#   Failed test 'Dumping (?^: \/ ) with PP'
#   at t/qr.t line 23.
# Global symbol "$VAR1" requires explicit package name (did you forget to declare "my $VAR1"?) at (eval 9) line 1.
# $VAR1 = qr/ \/ /;
# Looks like you failed 2 tests of 2.
../dist/Data-Dumper/t/qr.t .. 
1..2
not ok 1 - Dumping (?^: \/ ) with XS
not ok 2 - Dumping (?^: \/ ) with PP
Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/2 subtests 

Test Summary Report
-------------------
../dist/Data-Dumper/t/qr.t (Wstat: 512 Tests: 2 Failed: 2)
  Failed tests:  1-2
```
Likely cause:  The strings which `Data::Dumper::Dumper` writes are intended to be Perl data structures but are written as global variables.  If they are the `STRING` in `eval STRING`, the caller of the `eval` will register a violation of `strict 'vars'` and die.

@atoomic @toddr 

Thank you very much.
Jim Keenan

**Comment by atoomic** [comment]
this is now fixed as 3a7aea47

need review of these commits:
36c520e8b0
f9d7660e44
3a7aea47e3

**Comment by jkeenan** [comment]
> this is now fixed as [3a7aea4](https://github.com/atoomic/perl/commit/3a7aea47e3ebb69e15daa74678bd7ebfbcc3a318)
> 
> need review of these commits:
> [36c520e](https://github.com/atoomic/perl/commit/36c520e8b0bda712598e1ef7cd243b1f3a35e8d0)
> [f9d7660](https://github.com/atoomic/perl/commit/f9d7660e448adeb1120fdd4e4b66a26f6657f7ec)
> [3a7aea4](https://github.com/atoomic/perl/commit/3a7aea47e3ebb69e15daa74678bd7ebfbcc3a318)

Since our work here has a research/proof-of-concept focus, I think that the addition of this testing function:
```
sub add_my_to_dump {
    $_[0] =~ s{^(\s*)(\$VAR)}{$1 my $2}mg;

    return $_[0];
} 
```
... to the 3 remaining failing files is acceptable.  It demonstrates the problem.  But let's admit it -- it just gets us over until we can have a conversation with others about how Data-Dumper should evolve in future versions of Perl.

Thanks for your investigaton.
